tests/ops/test_channel_ops.py

```diff
@@ -213,6 +216,42 @@ def circuit(p):
         assert np.allclose(gradient, -(4 / 3) * np.cos(angle))
 
 
+class TestResetError:
+    """Tests for the quantum channel ResetError"""
+
+    @pytest.mark.parametrize("p_0,p_1", list(zip([0, 0.1, 0.5], [0, 0.1, 0.5])))
+    def test_p0_p1_arbitrary(self, p_0, p_1, tol):
+        """Test that various values of p_0 and p_1 give correct Kraus matrices"""
+        op = channel.ResetError
+
+        expected_K0 = np.sqrt(1 - p_0 - p_1) * np.eye(2)
+        assert np.allclose(op(p_0, p_1, wires=0).kraus_matrices[0], expected_K0, atol=tol, rtol=0)
+
+        expected_K1 = np.sqrt(p_0) * Zero
+        assert np.allclose(op(p_0, p_1, wires=0).kraus_matrices[1], expected_K1, atol=tol, rtol=0)
+
+        expected_K2 = np.sqrt(p_1) * One
+        assert np.allclose(op(p_0, p_1, wires=0).kraus_matrices[2], expected_K2, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("angle", np.linspace(0, 2 * np.pi, 7))
+    def test_grad_reset(self, angle, tol):
+        """Test that analytical gradient is computed correctly for different states. Channel
+        grad recipes are independent of channel parameter"""
+
+        dev = qml.device("default.mixed", wires=1)
+        p_0, p_1 = 0.5, 0.5
+
+        @qml.qnode(dev)
+        def circuit(p_0, p_1):
+            qml.RX(angle, wires=0)
+            qml.ResetError(p_0, p_1, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        gradient = np.squeeze(qml.grad(circuit)(p_0, p_1))
+        assert gradient == circuit(1) - circuit(0)
+        assert np.allclose(gradient, (-2 * np.cos(angle)))
```

Comment:
Great test @nahumsa! Did you calculate the expected gradient value, `-2 cos(angle)`, by hand?

Reply:
Yeah, I'm not sure about the analytical gradients for this one, I was using `TestBitFlip` as a reference. 
Maybe could you clarify a bit more about the gradient of this channel?

Reply:
One of the best sources for learning more about the analytic gradients of noisy channels is https://johannesjakobmeyer.com/blog/004-noisy-parameter-shift/, or alternatively, the PennyLane demonstration: https://pennylane.ai/qml/demos/tutorial_noisy_circuit_optimization.html

Reply:
In this particular case, I think @ixfoduap calculated the analytic value by hand, so hopefully he will be able to shed some light!

Reply:
Thanks for the followup. I'll read those.

Reply:
Hi @nahumsa, this is a great PR!

I'm not sure about the analytical gradients. The way we calculate analytical gradients is using this formula, which also appears in the blog post that Josh linked to:
![Screenshot from 2021-05-18 14-45-22](https://user-images.githubusercontent.com/40441298/118706804-cbd51580-b7e7-11eb-940d-03f8c9f461a6.png)

It is valid when the channel can be expressed as a linear combination of channels. This typically means that the Kraus matrices themselves should be a valid channel, which in practice often boils down to the Kraus matrices being unitaries.

In your case, the projectors `|0><0|` and `|1><1`| are not by themselves valid channels, so it's not accurate that the reset channel can be expressed as a linear combination of channels. This means that the analytical formula should not apply here. 

I believe that the reason the tests are passing is because you are setting both `p_0` and `p_1` equal to each other, which effectively makes this the identity channel --> `sqrt(p) |0><0| + sqrt(p) |1><1| = sqrt(p) (|0><0|+|1><1|) = sqrt(p) \identity` and we end up with the gradient of the gate in the circuit.

I would expect the test to fail if you make `p_0 != p_1` and compute the gradient by hand there. It's worth checking that this is the case. In any case, the gradient test should be checking gradients for many values of `p_0` and `p_1`.

If the analytical gradients do not apply here, it is not a big deal; it just means you should change the gradient method to "F" (finite difference) like it is done for other channels.

Reply:
Hi @ixfoduap! Thanks for the clarification. After reading the reference, I think that it is not possible to get analytic gradient by the formulas used on other channels.  
Thanks for the catch about the tests. After changing the gradient for using the finite difference methods should I delete the gradient test?

Reply:
Thanks @ixfoduap!

Reply:
@nahumsa let us know when you have the tests passing, or if you need a hand anywhere :)

Reply:
@josh146 and @ixfoduap, after revising some math, I found that I missed two Kraus operators that constitutes the `ResetError` channel! Sorry about that, I always find quantum channels quite tricky 😅